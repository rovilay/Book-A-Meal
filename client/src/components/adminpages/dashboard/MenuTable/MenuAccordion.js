import {
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody,
} from 'react-accessible-accordion';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import FontAwesome from 'react-fontawesome';
import moment from 'moment';
import swal from 'sweetalert';
import '../../../../assets/css/accordion.css';

class MenuAccordion extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isInfo: false,
    };

    this.handlePaginationClick = this.handlePaginationClick.bind(this);
  }

  /**
   * handles pagination changes
   *
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    const { Meals: mealUrl, } = this.props.item;
    const { limit } = this.props.menuMeals.pagination;
    const nextOffset = (data.selected) * limit;
    this.props.getMenuMeals(mealUrl, { limit, offset: nextOffset })
      .then(() => {
        this.setState({ isInfo: true });
      });
  }

  render() {
    const {
      isInfo,
    } = this.state;
    const {
      item,
      menuMeals,
      getMenuMeals,
      editMenu,
      deleteMenuMeal
    } = this.props;

    const { meals, pagination } = menuMeals;
    const {
      numOfPages,
      count
    } = pagination;

    const {
      sn,
      menuId,
      createdBy,
      Meals: mealUrl
    } = item;

    const postOn = moment(item.postOn).format('LL');
    const today = moment().format('YYYY-MM-DD');

    return (
      <AccordionItem key={sn}>
        <AccordionItemTitle>
          <h3>
            <span className="serial">{sn}</span>
            <span className="postOn">{postOn}</span>
          </h3>
          <FontAwesome
            name="chevron-down"
          />
        </AccordionItemTitle>

        <AccordionItemBody>
          <div className="accordion-body">
            <div className="menuContainer">
              <div className="menu-details">
                <p className="menu-id">
                  <span>Post On:</span>
                  <span>{postOn}</span>
                </p>
                <p className="menu-id">
                  <span>Created By:</span>
                  <span>{createdBy}</span>
                </p>
              </div>

              <div className="menu-meals">
                {
                  (!isInfo)
                  &&
                  (
                  <button
                    className="btn-2 show-meals-btn"
                    onClick={(e) => {
                      e.preventDefault();
                      getMenuMeals(mealUrl, {})
                        .then(() => {
                          this.setState({ isInfo: true });
                        });
                    }}
                  >
                    Show Meals
                  </button>
                  )
                }

                {
                  (isInfo) // for moble views
                  &&
                  (
                  <div className="meal-hide-add-btns show-mobile">
                    <button
                      className="responsive-btn-2 hide-meals-btn"
                      onClick={(e) => {
                        e.preventDefault();
                        this.setState({ isInfo: false });
                      }}
                    >
                      <FontAwesome
                        name = "eye-slash"
                        size = "lg"
                      />
                    </button>

                    {
                      // hide if menu date has passed
                      (moment(today).isSameOrBefore(item.postOn))
                      &&
                      <button
                        className="responsive-btn-2 add-meals-btn"
                        onClick={(e) => {
                          e.preventDefault();
                          editMenu({ menuId, postOn, meals });
                        }}
                      >
                        <FontAwesome
                          name = "plus"
                          size = "lg"
                        />
                      </button>
                    }
                  </div>
                  )
                }
                {
                  (isInfo) // for large screen
                  &&
                  (
                  <div className="meal-hide-add-btns hide-mobile">
                    <button
                      className="responsive-btn-2 hide-meals-btn"
                      onClick={(e) => {
                        e.preventDefault();
                        this.setState({ isInfo: false });
                      }}
                    >
                      Hide Meals
                    </button>
                    {
                      // hide if menu date has passed
                      (moment(today, 'YYYY-MM-DD').isSameOrBefore(item.postOn))
                      &&
                      <button
                        className="responsive-btn-2 add-meals-btn"
                        onClick={(e) => {
                          e.preventDefault();
                          editMenu({ menuId, postOn, meals });
                        }}
                      >
                        Add Meals
                      </button>
                    }
                  </div>
                  )
                }

                {
                  // menu meals
                  (meals.length > 0 && isInfo)
                  &&
                  <div className="meals">
                    {
                      meals.map((meal, i) => (
                        <div
                          id={meal.id}
                          className="meal"
                          key={i}
                        >
                          <span>
                            {meal.title}
                          </span>

                          {
                            // hide if menu date has passed
                            (moment(today, 'YYYY-MM-DD').isSameOrBefore(item.postOn))
                            &&
                            <button
                              className="btn-3 box-shadow"
                              onClick={(e) => {
                                e.preventDefault();

                                swal({
                                  text: 'Are you sure you want to remove this meal?',
                                  buttons: true,
                                  dangerMode: true,
                                })
                                  .then((confirmed) => {
                                    if (confirmed) {
                                      const newMenuMeals = meals.filter(n => n.id === meal.id);
                                      const MenuMealsToDelete = newMenuMeals.map(menuMeal => menuMeal.id);
                                      deleteMenuMeal({ mealUrl, meals: MenuMealsToDelete });
                                    }
                                  })
                                  .catch(err => err);
                              }}
                            >
                              <FontAwesome
                                name="trash"
                                size="2x"
                                className="trash"
                              />
                            </button>
                          }
                        </div>
                      ))
                    }
                  </div>
                }

                {
                  (meals.length < 1 && isInfo)
                  &&
                  <span className="empty not-found">
                      Your meals are not in this menu
                  </span>
                }
              </div>
              {
                (count > 5 && isInfo)
                &&
                <div className="pagination-container">
                  <ReactPaginate
                    previousLabel="<<"
                    nextLabel=">>"
                    breakLabel={<a href="">...</a>}
                    breakClassName="break-me"
                    pageCount={numOfPages}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={this.handlePaginationClick}
                    containerClassName="pagination"
                    subContainerClassName="pages pagination"
                    activeClassName="active"
                  />
                </div>
              }
            </div>
          </div>
        </AccordionItemBody>
      </AccordionItem>
    );
  }
}


MenuAccordion.propTypes = {
  item: PropTypes.object.isRequired,
  editMenu: PropTypes.func.isRequired,
  getMenuMeals: PropTypes.func.isRequired,
  deleteMenuMeal: PropTypes.func.isRequired,
  menuMeals: PropTypes.object.isRequired,
};

export default MenuAccordion;
