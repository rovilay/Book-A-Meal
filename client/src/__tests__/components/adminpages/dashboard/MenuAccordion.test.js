import React from 'react';
import { shallow } from 'enzyme';
import MenuAccordion from '../../../../components/adminpages/dashboard/MenuTable/MenuAccordion';
import { menus, menuMeals } from '../../../__mockData__/menuMock';

describe('MenuAccordion component test', () => {
  const setup = () => {
    const props = {
      item: menus[0],
      editMenu: jest.fn(),
      getMenuMeals: () => Promise.resolve(),
      deleteMenuMeal: jest.fn(),
      menuMeals: {
        meals: menuMeals,
        pagination: {
          limit: 12,
          count: 13,
          numOfPages: 2
        }
      },
    };

    return shallow(<MenuAccordion {...props} />)
  };


  it('should render MenuAccordion correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });


  it('should call `handlePaginationClick` menu meals pagination button is clicked',
  (done) => {
    const wrapper = setup();

    const data = {
      selected: 1
    };

    const handlePaginationClickSpy= jest.spyOn(wrapper.instance(), 'handlePaginationClick');
    wrapper.instance().handlePaginationClick(data)
    expect(handlePaginationClickSpy).toHaveBeenCalled();

    done();
  });


  it('should call `handleAccordionBodyClose` when menu accordion is clicked',
  (done) => {
    const wrapper = setup();

    document.body.innerHTML =
    `<div>
      <div type="checkbox" class="accordion" aria-selected="true" checked="true" />
    </div>`;

    const handleAccordionBodyCloseSpy= jest.spyOn(
      wrapper.instance(), 'handleAccordionBodyClose'
    );
    wrapper.instance().handleAccordionBodyClose('.accordion')
    expect(handleAccordionBodyCloseSpy).toHaveBeenCalled();

    done();
  });


  it('should call `handleAccordionItemTitleFocus` when menu accordion is on focus',
  (done) => {
    const wrapper = setup();

    const handleAccordionItemTitleFocusSpy= jest.spyOn(
      wrapper.instance(), 'handleAccordionItemTitleFocus'
    );
    wrapper.instance().handleAccordionItemTitleFocus()
    expect(handleAccordionItemTitleFocusSpy).toHaveBeenCalled();

    done();
  });


  it('should call `showMenuMeals` if show meals button is clicked', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn()
    }

    const showMenuMealsSpy= jest.spyOn(wrapper.instance(), 'showMenuMeals');
    wrapper.instance().showMenuMeals(event)
    expect(showMenuMealsSpy).toHaveBeenCalled();

    done();
  });


  it('should call `hideMenuMeals` if hide meals button is clicked', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn()
    }

    const hideMenuMealsSpy= jest.spyOn(wrapper.instance(), 'hideMenuMeals');
    wrapper.instance().hideMenuMeals(event)
    expect(hideMenuMealsSpy).toHaveBeenCalled();

    done();
  });


  it('should call `handleAddMealsToMenu` if add meal button is clicked',
  (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn()
    }

    const handleAddMealsToMenuSpy= jest.spyOn(wrapper.instance(), 'handleAddMealsToMenu');
    wrapper.instance().handleAddMealsToMenu(event)
    expect(handleAddMealsToMenuSpy).toHaveBeenCalled();

    done();
  });


  it('should call `deleteMealInMenu` if delete meal button is clicked', (done) => {
    const wrapper = setup();

    const deleteMealInMenuSpy= jest.spyOn(wrapper.instance(), 'deleteMealInMenu');
    wrapper.instance().deleteMealInMenu(menuMeals[0])
    expect(deleteMealInMenuSpy).toHaveBeenCalled();

    done();
  });

  it('should `deleteMealInMenu` if delete button is clicked', (done) => {
    const wrapper = setup();
    wrapper.setState({
      isInfo: true
    });

    const event = {
      preventDefault: () => jest.fn()
    }

    wrapper.find('.deleteMeal').first().simulate('click', event);
    const deleteMealInMenuSpy= jest.spyOn(wrapper.instance(), 'deleteMealInMenu');
    wrapper.instance().deleteMealInMenu(menuMeals[0])
    expect(deleteMealInMenuSpy).toHaveBeenCalled();

    done();
  });
});
