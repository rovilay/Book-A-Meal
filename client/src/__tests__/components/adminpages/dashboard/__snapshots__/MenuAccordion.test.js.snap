// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuAccordion component test should render MenuAccordion correctly 1`] = `
<AccordionItemWrapper
  accordionStore={
    AccordionContainer {
      "_listeners": Array [],
      "addItem": [Function],
      "removeItem": [Function],
      "setAccordion": [Function],
      "setExpanded": [Function],
      "setOnChange": [Function],
      "state": Object {
        "accordion": true,
        "items": Array [],
        "onChange": [Function],
      },
    }
  }
  className="accordion__item"
  disabled={false}
  expanded={false}
  hiddenbodyclassname="myhide-accordion"
  hideBodyClassName=""
>
  <AccordionItemTitleWrapper
    className="accordion__title myaccordiontitle-undefined"
    hideBodyClassName=""
    onFocus={[Function]}
  >
    <h3>
      <span
        className="serial"
      />
      <span
        className="postOn"
      >
        September 10, 2018
      </span>
    </h3>
    <FontAwesome
      name="chevron-down"
    />
  </AccordionItemTitleWrapper>
  <AccordionItemBodyWrapper
    className="accordion__body"
    hideBodyClassName="accordion__body--hidden"
  >
    <div
      className="accordion-body"
    >
      <div
        className="menuContainer"
      >
        <div
          className="menu-details"
        >
          <p
            className="menu-id"
          >
            <span>
              Post On:
            </span>
            <span>
              September 10, 2018
            </span>
          </p>
          <p
            className="menu-id"
          >
            <span>
              Created By:
            </span>
            <span />
          </p>
        </div>
        <div
          className="menu-meals"
        >
          <button
            className="btn-2 show-meals-btn"
            onClick={[Function]}
          >
            Show Meals
          </button>
        </div>
      </div>
    </div>
  </AccordionItemBodyWrapper>
</AccordionItemWrapper>
`;
